server/bleep/src/webserver/query.rs

```diff
@@ -85,112 +86,32 @@ impl ApiQuery {
-                indexes.file.count(queries.iter(), &FileReader),
-            )
-            .map_err(|e| EndpointError {
-                kind: ErrorKind::Internal,
-                message: e.to_string().into(),
-            })?;
-
-            // total number of results is the sum of total results per index
-            let total_count = total_content_count + total_repo_count + total_file_count;
-
-            // total pages is the number of pages required to accommodate the greatest
-            // of the three results
-            let page_count = {
-                let max_items = total_content_count
-                    .max(total_repo_count)
-                    .max(total_file_count);
-                div_ceil(max_items, self.page_size)
-            };
-
-            (Some(page_count), Some(total_count))
-        } else {
-            (None, None)
-        };
-
-        let metadata = QueryMetadata {
-            page: self.page,
-            page_size: self.page_size,
-            page_count,
-            total_count,
-        };
-
-        let data = query_result_list;
-
-        let response = QueryResponse {
-            count,
-            data,
-            stats,
-            metadata,
-        };
-
-        Ok(response)
-    }
-}
-
-#[derive(Debug, Default, Serialize, Deserialize)]
-pub struct ResultStats {
-    pub repo: HashMap<String, usize>,
-    pub lang: HashMap<String, usize>,
-}
-
-fn gather_result_stats(query_results: &[QueryResult]) -> ResultStats {
-    let mut stats = ResultStats::default();
-    for result in query_results.iter() {
-        match result {
-            QueryResult::Snippets(data) => {
-                *stats.repo.entry(data.repo_name.clone()).or_insert(0) += 1;
-                if let Some(l) = &data.lang {
-                    *stats.lang.entry(l.clone()).or_insert(0) += 1;
-                }
+        for q in &queries {
+            if ContentReader.query_matches(q) {
+                return ContentReader
+                    .execute(&indexes.file, &queries, &self)
+                    .await
+                    .map_err(|e| EndpointError::internal(e.to_string().into()));
+            } else if RepoReader.query_matches(q) {
+                return RepoReader
+                    .execute(&indexes.repo, &queries, &self)
+                    .await
+                    .map_err(|e| EndpointError::internal(e.to_string().into()));
+            } else if FileReader.query_matches(q) {
+                return FileReader
+                    .execute(&indexes.file, &queries, &self)
+                    .await
+                    .map_err(|e| EndpointError::internal(e.to_string().into()));
+            } else if OpenReader.query_matches(q) {
+                return OpenReader
+                    .execute(&indexes.file, &queries, &self)
+                    .await
+                    .map_err(|e| EndpointError::internal(e.to_string().into()));
```

Comment:
Looks like this change breaks queries with multiple different types, e.g. `foobar or symbol:foo`. Can we revert to the old behaviour here?

Reply:
Makes sense. This has the side effect of interpreting `repo:foo or bar` as a repo-only query. Can you add a comment above this condition section for future clarity as to what's happening?